packages/webview_flutter/webview_flutter_platform_interface/lib/src/v4/navigation_callback_handler_delegate.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:plugin_platform_interface/plugin_platform_interface.dart';
+
+import 'types/types.dart';
+import 'webview_platform.dart';
+
+/// Interface for callbacks made by [NavigationCallbackHandlerDelegate].
+///
+/// The webview plugin implements this class, and passes an instance to the
+/// [NavigationCallbackHandlerDelegate].
+/// [NavigationCallbackHandlerDelegate] is notifying this handler on events that
+/// happened on the platform's webview.
+abstract class NavigationCallbackHandlerDelegate extends PlatformInterface {
+  /// Creates a new [NavigationCallbacksHandlerDelegate]
+  factory NavigationCallbackHandlerDelegate() {
+    final NavigationCallbackHandlerDelegate callbackHandlerDelegate =
+        WebViewPlatform.instance!.createNavigationCallbackHandlerDelegate();
+    PlatformInterface.verify(callbackHandlerDelegate, _token);
+    return callbackHandlerDelegate;
+  }
+
+  /// Used by the platform implementation to create a new
+  /// [NavigationCallbackHandlerDelegate].
+  ///
+  /// Should only be used by platform implementations because they can't extend
+  /// a class that only contains a factory constructor.
+  @protected
+  NavigationCallbackHandlerDelegate.implementation() : super(token: _token);
+
+  static final Object _token = Object();
+
+  /// Invoked by [WebViewPlatformControllerDelegate] when a navigation request
+  /// is pending.
+  ///
+  /// If true is returned the navigation is allowed, otherwise it is blocked.
+  FutureOr<bool> onNavigationRequest(
```

Comment:
We don't want any of the callback methods to be required to implement. This one can return `true` by default and the rest should have empty implementations.

Reply:
Actually I think we want these to take an anonymous function. As in 

```dart
Future<void> setOnNavigationRequest(void Function({required String url, required bool isForMainFrame}) onNavigationRequest) {
}
```

I think this makes it easier for the app-facing interface and the implementations.

Reply:
@bparrishMines this makes sense, do I implement it in the platform interface to throw a not implemented error or should I also create backing fields and set those?

If we go for the later, should I also implement getter methods and if so shouldn't we skip the separate `set...` and `get..` methods and replace them by public fields?

Reply:
Yea, it should throw an `UnimplmentedError()`. For now, I think we only need the setter methods.

We can't use fields because the implementations would break the style guides for getters and setters when they use them to make a `MethodCall`. 
Style guide for getters: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#getters-feel-faster-than-methods
Style guide for setters: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#perform-dirty-checks-in-setters

I recently had to write PR to fix me making the same mistake.

Reply:
Converted all methods to `setOn...` and to accept an anonymous function as parameter. Also added additional tests to ensure the default interface throws an `UnimplementedError`.